### PR TITLE
Laboratory work #3, Ugolnikov Viacheslav - 22FPL1

### PR DESCRIPTION
**Pull request by slava-ugolnikov**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -16,13 +16,15 @@ class TextProcessor:
         _storage (dict): Dictionary in the form of <token: identifier>
     """
 
-    def __init__(self, end_of_word_token: str) -> None:
+    def __init__(self, end_of_word_token: str, storage: dict) -> None:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:19]: Do not change the interface

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         """
+        if not isinstance(text, str) or not text:
+            return None
+
+        tokenized_text = []
+
+        for symbol in text.lower():
+            if symbol.isalpha():
+                tokenized_text.append(symbol)
+            elif symbol in ' !?.' \
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:54]: Only these symbols are punctuation?

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -159,6 +241,25 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
         """
+        if not isinstance(decoded_corpus, tuple) \
+                or not decoded_corpus:
+            return None
+
+        decoding_list = []
+
+        for token in list(decoded_corpus):
+            if token == self._end_of_word_token:
+                decoding_list += ' '
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:252]: again. It is a workaround with magic constants. It is a bad solution

File: seminars/practice_8_inheritance.py [hunk 4]
```diff
@@ -11,8 +11,14 @@
     # Methods:
         # move
 
+
 class Vehicle:
-    ...
+    def __init__(self, max_speed: int, colour: str) -> None:
```
Comment (artyomtugaryov) [inline seminars/practice_8_inheritance.py:16]: checkout to original version

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -41,6 +43,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         """
+        if not isinstance(text, str) or not text:
+            return None
+
+        tokenized_text = []
+
+        for symbol in text.lower():
+            if symbol.isalpha():
+                tokenized_text.append(symbol)
+            elif symbol in '''!"#$%&'()*+, ./\:;<=>?@[]^_`{|}~''' \
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:54]: Is it all punctuation? Is there reliable way to get punctuation?

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -76,6 +100,15 @@ def get_token(self, element_id: int) -> Optional[str]:
...
         """
+        if not (isinstance(element_id, int)
+                and element_id in self._storage.values()):
+            return None
+
+        for key, value in self._storage.items():
+            if value == element_id:
+                return key
+
+        return None
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:111]: There is a better way to fix mypy/pylint...

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -93,6 +126,25 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
+        if not tokenized_text:
+            return None
+
+        encoded_corpus = []
+
+        for symbol in tokenized_text:
+            self._put(symbol)
+            encoded_corpus.append(self.get_id(symbol))
+
+        if None in tokenized_text \
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:142]: can you check it inside loop and return None as soon as posible

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -104,6 +156,12 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not (isinstance(element, str)
+                and len(element) == 1):
+            return None
+
+        if element not in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:163]: what about early return?

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -159,6 +243,21 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+
+        for token in list(decoded_corpus):
+            decoding_list.append(token)
+
+        if decoding_list[-1] == self._end_of_word_token:
+            decoding_list.pop()
+
+        result = ''.join(decoding_list).capitalize().replace(self._end_of_word_token, ' ')
+
+        return result + '.'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:260]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -500,3 +647,4 @@ def _get_next_token(self, sequence_to_continue: tuple[int, ...]) -> Optional[dic
 
         In case of corrupt input arguments return None.
         """
+
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:650]: revert

File: seminars/practice_8_inheritance.py (hunk 4 above)
Comment (artyomtugaryov) [reply seminars/practice_8_inheritance.py:16]: !

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: seminars/practice_8_inheritance.py (hunk 4 above)
Comment (slava-ugolnikov) [reply seminars/practice_8_inheritance.py:16]: how can I do that?

File: lab_3_generate_by_ngrams/main.py (hunk 8 above)
Comment (slava-ugolnikov) [reply lab_3_generate_by_ngrams/main.py:163]: I tried to do that

File: lab_3_generate_by_ngrams/main.py (hunk 9 above)
Comment (slava-ugolnikov) [reply lab_3_generate_by_ngrams/main.py:260]: done

**Pull request closed without merging**